repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -0,0 +1,186 @@
+
+        } else {
+            String vertexName = vertex.getProperty(NAME, String.class);
+            if (!vertexName.equals(productName)) {
+                domainExists(productName, newDomainQualifiedName);
+            }
+
+            entity.setAttribute(QUALIFIED_NAME, vertexQnName);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void processMoveDataProductToAnotherDomain(AtlasEntity product,
+                                                     String sourceDomainQualifiedName,
+                                                     String targetDomainQualifiedName,
+                                                     String currentDataProductQualifiedName) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder recorder = RequestContext.get().startMetricRecord("processMoveDataProductToAnotherDomain");
+
+        try {
+            String domainName = (String) product.getAttribute(NAME);
+
+            LOG.info("Moving dataProduct {} to Domain {}", domainName, targetDomainQualifiedName);
+
+            domainExists(domainName, targetDomainQualifiedName);
+
+            String updatedQualifiedName = currentDataProductQualifiedName.replace(sourceDomainQualifiedName, targetDomainQualifiedName);
+
+            product.setAttribute(QUALIFIED_NAME, updatedQualifiedName);
+
+            LOG.info("Moved dataProduct {} to Domain {}", domainName, targetDomainQualifiedName);
+
+        } finally {
+            RequestContext.get().endMetricRecord(recorder);
+        }
+    }
+
+    private void setParent(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("DataProductPreProcessor.setParent");
+        if (parentDomain == null) {
+            AtlasObjectId objectId = (AtlasObjectId) entity.getRelationshipAttribute(PARENT_DOMAIN);
+
+            if (objectId != null) {
+                if (StringUtils.isNotEmpty(objectId.getGuid())) {
+                    AtlasVertex vertex = context.getVertex(objectId.getGuid());
+
+                    if (vertex == null) {
+                        parentDomain = entityRetriever.toAtlasEntityHeader(objectId.getGuid());
+                    } else {
+                        parentDomain = entityRetriever.toAtlasEntityHeader(vertex);
+                    }
+
+                } else if (MapUtils.isNotEmpty(objectId.getUniqueAttributes()) &&
+                        StringUtils.isNotEmpty((String) objectId.getUniqueAttributes().get(QUALIFIED_NAME))) {
+                    parentDomain = new AtlasEntityHeader(objectId.getTypeName(), objectId.getUniqueAttributes());
+
+                }
+            }
+        }
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void domainExists(String domainName, String domainQualifiedName) throws AtlasBaseException {
```

Comment:
Why do we need this call for Domains? do we want to restrict moving Domains with same name?

Reply:
This is done to avoid more than one subdomains with same name inside a Domain after move